### PR TITLE
migrate compiler, bootstrap and compiletest to windows-rs

### PR DESCRIPTION
**Pull request by euclio**
This PR migrates the compiler, bootstrap, and compiletest to use [windows-rs](https://github.com/microsoft/windows-rs) instead of winapi-rs. windows-rs is the bindings crate provided by Microsoft, and is actively maintained compared to winapi-rs. Not all ecosystem crates have migrated over yet, so there will be a period of time where both crates are used.

windows-rs also provides some nice ergonomics over winapi-rs to convert return values to `Result`s (which found a case where we forgot to check the return value of `CreateFileW`).

### CODE REVIEW

**Comment by klensy** [comment]
Shouldn't windows-sys be used instead?

https://github.com/microsoft/windows-rs#windows-sys: `The windows-sys crate is a zero-overhead fallback for the most demanding situations`

**Comment by ChrisDenton** [comment]
Unless it makes compiling rustc significantly slower, I think the `windows` crate is better for this. The "overhead", such as returning `Result`s, is helpful and is something we'd want to do anyway.

**Comment by klensy** [comment]
>Unless it makes compiling rustc significantly slower

But there no windows perf runs, yet.

**Comment by ChrisDenton** [comment]
Sure. I wouldn't expect the compile time of one crate to have a significant impact here though. And if it does we should see slower CI times. We can always switch to windows-sys if it does cause problems.

**Review by ChrisDenton — COMMENTED** [review]
Looks good. I just have a few comments, mostly nits.

File: compiler/rustc_errors/src/lock.rs
Comment (klensy) [inline compiler/rustc_errors/src/lock.rs:53]: This will unconditionally allocate error string, `unwrap_or_else( || panic! ...)` should do the thing.

**Comment by jyn514** [comment]
The idea seems fine, but I'm short on review time and not familiar with Windows.

r? @ChrisDenton (thank you for your reviews so far ❤️)

File: compiler/rustc_data_structures/src/profiling.rs
Comment (euclio) [reply compiler/rustc_data_structures/src/profiling.rs]: Fixed.

File: compiler/rustc_data_structures/src/profiling.rs
Comment (euclio) [reply compiler/rustc_data_structures/src/profiling.rs]: Fixed.

File: compiler/rustc_errors/src/lock.rs
Comment (euclio) [reply compiler/rustc_errors/src/lock.rs]: Fixed.

File: compiler/rustc_session/src/filesearch.rs
Comment (euclio) [reply compiler/rustc_session/src/filesearch.rs]: Fixed.

File: src/bootstrap/bin/rustc.rs
Comment (euclio) [reply src/bootstrap/bin/rustc.rs:279]: Personally I prefer limiting the scope as much as possible, to just the FFI calls.

File: src/bootstrap/util.rs
Comment (euclio) [reply src/bootstrap/util.rs]: Fixed.

File: compiler/rustc_errors/src/lock.rs
Comment (euclio) [reply compiler/rustc_errors/src/lock.rs]: Fixed.

**Comment by euclio** [comment]
Sorry for the ping, accidentally included the cargo submodule in the previous rebase.

**Comment by ChrisDenton** [comment]
I'm happy to review this; I'd mostly be keen to make sure the API calls end up the same. However, I'm not sure that I'm actually authorized to sign off on this change (I'm more std/libs). cc @Mark-Simulacrum maybe?

Moving on from winapi (which is no longer updated) makes sense to do at some point. @klensy brings up the point that maybe we should be using `windows-sys` instead of `windows` due to compile times (see [comment](https://github.com/rust-lang/rust/pull/106610#issuecomment-1375475767)). But I would be keen to at least try `windows` for correctness reasons (e.g. its use of `Result` helps avoid at least one source of potential mistakes).

**Comment by jyn514** [comment]
@ChrisDenton consider this permission to sign off of the change :) I trust your judgement.

@bors delegate=ChrisDenton

**Comment by ChrisDenton** [comment]
Oh ok! In that case I'm fine with this as long as CI is also ok.

@bors r+ rollup=never

**Comment by euclio** [comment]
Hm, that's a weird failure. I can fix it by adding a `windows-rs` dependency to librustdoc, but that seems a bit hacky since it doesn't depend on it directly.

**Comment by jyn514** [comment]
cc @bjorn3, I think I remember you saying on Zulip that linkage metadata wasn't propagated through rlibs or something like that. Don't recall the thread off the top of my head.

**Comment by bjorn3** [comment]
> This is empty but its build script adds the necessary native library path. Presumably this is not passed on when using extern crate to include rlibs because the crate is using cargo build scripts which rustc itself isn't aware of.

Indeed. This will be a problem with every tool that uses rustc_driver, not just rustdoc. I think we would need to copy `windows.lib` to the sysroot to allow to allow the linker to find it when building external tools. The sysroot is already in the linker search path for libstd.so and librustc_driver.so. The alternative solution is for windows-rs to switch to raw-dylib. That way windows.lib isn't necessary, but instead will be generated by rustc and stored in the rlib. I'm not sure if that breaks due to my recent change to only ship rmeta and dylib files and not rlib files though.

> cc @bjorn3, I think I remember you saying on Zulip that linkage metadata wasn't propagated through rlibs or something like that. Don't recall the thread off the top of my head.

I think that was when responding to a Buck developer. They are renaming .rlib to .a, which loses the entire crate metadata, not just linking information. At least library names are stored in the crate metadata. **That said the library search path is indeed not stored, which is the problem here.** However we can't change it to be stored as that did completely break reproducible builds.

**Comment by ChrisDenton** [comment]
> I think we would need to copy windows.lib to the sysroot to allow to allow the linker to find it when building external tools.

Ok, so this looks like the best way forward atm. But in the future we may move to using raw_dylib, assuming that avoids the issue.

**Comment by euclio** [comment]
Sounds good, what would be the correct way to retrieve and copy it into the sysroot?

**Comment by jyn514** [comment]
@euclio I'm not sure how to find the file in cargo's target directory (maybe you can use the JSON output and filter for artifacts named `windows.lib`?) but you can copy it to the sysroot like so:
https://github.com/rust-lang/rust/blob/89ccd704f04f86e70a58e8ae6867656503e4b89c/src/bootstrap/compile.rs#L897-L912

**Comment by jyn514** [comment]
ok actually this looks more relevant: https://github.com/rust-lang/rust/blob/89ccd704f04f86e70a58e8ae6867656503e4b89c/src/bootstrap/compile.rs#L1443-L1451
I don't think we should be skipping `.lib` files, at least not unconditionally.

**Comment by bjorn3** [comment]
Not skipping .lib files there won't have any effect. The windows.lib file never ends up in the deps/ dir from which we copy to the sysroot. It stays in the source dir of the respective crate. I think we could look in the build/ dir for the respective crate and parse the output of the build script stored there to get the link path from which we can copy windows.lib into the sysroot.

**Comment by bjorn3** [comment]
Also there is an inversion a couple lines up. Meaning that this code actually makes it such that .lib files are **not** skipped.

**Comment by euclio** [comment]
Pushed a new commit that enables `raw-dylib` support in window-rs via a `cfg` introduced in 0.44.

**Comment by mati865** [comment]
Raw dylib might fall on `i686-pc-windows-gnu`: https://github.com/rust-lang/rust/issues/104453

**Comment by SmiteWindows** [comment]
0.45 try

**Comment by euclio** [comment]
> Raw dylib might fall on `i686-pc-windows-gnu`: #104453

The linked issue is now fixed.

**Comment by kennykerr** [comment]
There might still be a compiler issue with `raw-dylib` support. I can't get it to work reliably with `i686-pc-windows-gnu`.

https://github.com/microsoft/windows-rs/pull/2312

File: src/bootstrap/job.rs
Comment (kennykerr) [reply src/bootstrap/job.rs:140]: `.ok()` above turns it into a `Result`.

Comment (riverar) [reply src/bootstrap/job.rs:140]: Thanks, missed that!

**Comment by kennykerr** [comment]
Update: false alarm - @riverar figured out that I was hitting into a flaky gnu setup on GitHub (https://github.com/microsoft/windows-rs/pull/2312). So you should be good to go.

**Review by riverar — APPROVED** [review]

**Comment by ChrisDenton** [comment]
Using raw_dylib seems good to me. Let's try this again...

@bors r+

**Comment by riverar** [comment]
I was able to compile that branch locally and get through `tests\ui-fulldeps` with no issues using external msys/mingw-w64-x86_64.

So where do we go from here? Is there perhaps an issue with the bootstrapped mingw bits? Is that something I can bring down locally?

**Comment by ChrisDenton** [comment]
You can download the CI mingw from here: https://ci-mirrors.rust-lang.org/rustc/x86_64-12.2.0-release-posix-seh-rt_v10-rev0.7z. It's an older version due to Rust still needing to target Windows 7.

**Comment by riverar** [comment]
GNU dlltool, as part of its import library generation logic, uses a bunch of temporary files on disk. In the interest of deterministic build runs, dlltool supports deterministic temporary filenames. The temporary filename prefix is automatically generated internally or can be explicitly specified via a `--temp-prefix` argument.

GNU dlltool **2.38** (that ships with `x86_64-12.2.0-release-posix-seh-rt_v10-rev0` [installed during CI](https://github.com/rust-lang/rust/blob/master/src/ci/scripts/install-mingw.sh)) generates a prefix based on the target library name ([source](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=binutils/dlltool.c;h=d95bf3f5470b999fa3b30bc887791859f48d81d1;hb=20756b0fbe065a84710aa38f2457563b57546440#l3992)). The tool writes to files such as `target_dll_h.s` and `target_dll_s00203.o`.

This presents a problem when multiple instances of rustc_codegen_llvm are used to generate an import library (as part of the raw_dylib feature) for the same target library (e.g. kernel32) ([source](https://github.com/rust-lang/rust/blob/master/compiler/rustc_codegen_llvm/src/back/archive.rs#L185-L196)). That is, dlltool instances race and may overwrite or delete files belonging to each other.

Re-running the failing tests in CI with `-j1` allievates the issue.

GNU dlltool **2.39**+ (not used in Rust CI) generates a prefix based on the output library path ([source](https://sourceware.org/git/?p=binutils-gdb.git;a=blob;f=binutils/dlltool.c;h=e2af20847009945b4c61a6fef08268fbb4429715;hb=b51c2fec1da205ea3e7354cbb3e253018d64873c#l3992)). The tool, when invoked as part of a Rust compile pass, writes to files such as `C_Users_Foo_AppData_Local_Temp_rustcOFqhXZ_target_lib_h.s`. (The output library path is normalized and non-alphanumeric characters are replaced with underscores.)

Updating dlltool in CI incurs a high risk of breaking something else. Instead, I recommend we alter rustc_codegen_llvm to explicitly specify a temporary filename prefix and emulate what GNU dlltool 2.39 does to ensure an eventual dlltool upgrade goes smoothly.

I believe something like this should do the trick:
```diff
let result = std::process::Command::new(dlltool)
    .args([
        "-d",
        def_file_path.to_str().unwrap(),
        "-D",
        lib_name,
        "-l",
        output_path.to_str().unwrap(),
        "--no-leading-underscore",
+     "--temp-prefix"
+     output_path.parent().unwrap()
    ])
    .output();
```

**Comment by riverar** [comment]
@euclio One more merge to catch up then we can retry a build.

**Comment by mati865** [comment]
Isn't bootstrap built with beta compiler?
In that case this would have to wait util around half of March.

**Comment by riverar** [comment]
@mati865 Ah right, oops, I'm still working my way through the rustc book. Thanks!

**Comment by klensy** [comment]
> Isn't bootstrap built with beta compiler? In that case this would have to wait util around half of March.

PR can be split into separate parts and merged. How bootstrap prevents this?

**Comment by riverar** [comment]
> > Isn't bootstrap built with beta compiler? In that case this would have to wait util around half of March.
> 
> PR can be split into separate parts and merged. How bootstrap prevents this?

Think we're just waiting for bootstrap bits to get built and used during CI. This PR fails without https://github.com/rust-lang/rust/pull/107752

**Comment by weihanglo** [comment]
Cargo bump (#108665) also suffers from a similar issue. The PR updates `tempfile` dependency which switches to `windows-sys`. Tests `ui-fulldeps` [started failing on windows-msvc](https://github.com/rust-lang-ci/rust/actions/runs/3932834003/jobs/6725872323). I will really appreciate if anyone gives me a help to unblock cargo update 🙇🏾.

**Comment by mati865** [comment]
https://github.com/rust-lang/rust/pull/108665 looks unrelated since it fails on `windows-msvc` target while this PR in blocked on `windows-gnu` target.

**Comment by klensy** [comment]
Bootstrap compiler was updated.

**Comment by riverar** [comment]
@euclio Hey Andy, mind resolving the conflict here so we can give bors another whack at this?

**Comment by kennykerr** [comment]
Sweet - thanks all - that was some serious collaboration. 🎉

**Comment by ChrisDenton** [comment]
Yay! 🎉

It's in nightly now so we'll see how this goes...

**Comment by klensy** [comment]
Hm, export table size of rustc_driver changed from 18104 to 25790 items (some nearby nightly), plus it looks like there imported few unused fn from windows libs.

**Comment by ChrisDenton** [comment]
Just for future reference, the two relevant nightly versions are:

rustc 1.70.0-nightly (44f518058 2023-03-20)
rustc 1.70.0-nightly (da7c50c08 2023-03-19)

Though in rustup you'll need to +1 to those dates. E.g.:

```
> rustup toolchain install nightly-2023-03-20-x86_64-pc-windows-msvc
> rustup toolchain install nightly-2023-03-21-x86_64-pc-windows-msvc
```

**Comment by kennykerr** [comment]
Thanks, yeah if it's a lib then that's not too surprising.

**Comment by dpaoliello** [comment]
> Thanks, yeah if it's a lib then that's not too surprising.

It's not a lib, it's a dll that is exporting Rust functions (rather than C functions). If I understand Rust's model for this, then all `pub` Rust functions must be exported from that dll in case a dependent of `rustc_driver` attempts to use a function in one `rustc_driver`'s dependencies.

Would it be possible to trim down the API exposed by `windows-rs` to reduce the number of `pub` functions?

**Comment by kennykerr** [comment]
> Would it be possible to trim down the API exposed by windows-rs to reduce the number of pub functions?

Just using Cargo features - [there's a discussion here](https://github.com/microsoft/win32metadata/issues/1500) about reducing the size of the largest namespaces where I mentioned the `Windows.Win32.System.Diagnostics.Debug` namespace listed in your dump.

Reducing the size of that namespace would go a long way toward reducing the number of `pub` functions in the scenario.

**Comment by kennykerr** [comment]
> Would it be possible to trim down the API exposed by windows-rs to reduce the number of pub functions?

https://github.com/microsoft/windows-rs/releases/tag/0.47.0 has been published and includes an update to the metadata resulting in a much smaller `Win32_System_Diagnostics_Debug` feature. This should go a long way toward reducing the number of `pub` functions.

**Pull request merged**
